app/controllers/application_controller.rb

```diff
@@ -1541,6 +1541,22 @@ def show_index_of_objects(query, args = {})
 
   private ##########
 
+  class << self
+    attr :index_subaction_dispatch_table, :index_subaction_param_keys
+  end
+
+  def dispatch_to_index_subaction
+    self.class.index_subaction_param_keys.each do |subaction|
+      if params[subaction].present?
+        return send(
+          self.class.index_subaction_dispatch_table[subaction] ||
+          subaction
+        )
+      end
+    end
+    default_index_subaction
+  end
```

Comment:
@mo-nathan:  Does this look OK?
As you suggested, things break if the key list and the dictionary are constants in the subclasses. 
This is my attempt at instead using class instance variables.
I could use constants if you don't mind including them as arguments. That would also deal with another issue.  See the comment on app/controllers/observations_controller/index.rb

Reply:
What you have here looks right to me.